File: src/resolvePlaywright.ts
import { existsSync } from 'node:fs';
import path from 'node:path';

export interface ResolveOptions {
  readonly cwd?: string;
  readonly explicitBin?: string;
  readonly exists?: (filePath: string) => boolean;
  readonly nodePath?: string;
  readonly platform?: NodeJS.Platform;
}

export interface PlaywrightInvocation {
  bin: string;
  args: string[];
}

const GLOBAL_PLAYWRIGHT_BIN = 'playwright';

// Paths are relative to node_modules, in order of preference.
const WINDOWS_CLI_SCRIPTS: readonly (readonly string[])[] = [
  ['playwright', 'cli.js'],
  ['@playwright', 'test', 'cli.js'],
];
const POSIX_BIN: readonly string[] = ['.bin', 'playwright'];

export function resolvePlaywrightBin(options: ResolveOptions = {}): string {
  return resolvePlaywrightInvocation(options).bin;
}

export function resolvePlaywrightInvocation(options: ResolveOptions = {}): PlaywrightInvocation {
  const {
    cwd = process.cwd(),
    explicitBin,
    exists = existsSync,
    nodePath = process.execPath,
    platform = process.platform,
  } = options;

  if (explicitBin) return { bin: explicitBin, args: [] };

  const nodeModules = path.join(cwd, 'node_modules');

  if (platform === 'win32') {
    // The .cmd shim in node_modules/.bin needs a shell, so run the CLI script through node instead.
    const cliScript = WINDOWS_CLI_SCRIPTS.map((segments) => path.join(nodeModules, ...segments)).find((candidate) =>
      exists(candidate),
    );
    if (cliScript) return { bin: nodePath, args: [cliScript] };

    return { bin: GLOBAL_PLAYWRIGHT_BIN, args: [] };
  }

  const localBin = path.join(nodeModules, ...POSIX_BIN);
  if (exists(localBin)) return { bin: localBin, args: [] };

  return { bin: GLOBAL_PLAYWRIGHT_BIN, args: [] };
}

File: src/resolve_playwright.rs
//! Port of `resolvePlaywright.ts`.

use std::path::{Path, PathBuf};

#[derive(Default)]
pub struct ResolveOptions<'a> {
    pub cwd: Option<&'a str>,
    pub explicit_bin: Option<&'a str>,
    pub exists: Option<&'a dyn Fn(&str) -> bool>,
    pub node_path: Option<&'a str>,
    pub platform: Option<&'a str>,
}

#[derive(Debug, PartialEq)]
pub struct PlaywrightInvocation {
    pub bin: String,
    pub args: Vec<String>,
}

pub fn resolve_playwright_bin(options: &ResolveOptions) -> String {
    resolve_playwright_invocation(options).bin
}

pub fn resolve_playwright_invocation(options: &ResolveOptions) -> PlaywrightInvocation {
    if let Some(bin) = options.explicit_bin {
        return invocation(bin, vec![]);
    }

    let cwd = options
        .cwd
        .map(PathBuf::from)
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_default());
    let default_exists = |path: &str| Path::new(path).exists();
    let exists = options.exists.unwrap_or(&default_exists);
    let platform = options
        .platform
        .unwrap_or(if cfg!(windows) { "win32" } else { "unix" });

    if platform == "win32" {
        let node = options.node_path.unwrap_or("node");
        for cli in ["node_modules/playwright/cli.js", "node_modules/@playwright/test/cli.js"] {
            let local_cli = join(&cwd, cli);
            if exists(&local_cli) {
                return invocation(node, vec![local_cli]);
            }
        }
        return invocation("playwright", vec![]);
    }

    let local_bin = join(&cwd, "node_modules/.bin/playwright");
    if exists(&local_bin) {
        return invocation(&local_bin, vec![]);
    }

    invocation("playwright", vec![])
}

fn invocation(bin: &str, args: Vec<String>) -> PlaywrightInvocation {
    PlaywrightInvocation { bin: bin.to_string(), args }
}

fn join(cwd: &Path, relative: &str) -> String {
    cwd.join(relative).to_string_lossy().into_owned()
}
